### PR TITLE
fix: add select options field to new row

### LINKED ISSUE #20230: [Bug]-[400]:For select column type in the table widget where the options are dynamically set, the same doesn't show up during Add new row flow
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

For select column type in the table widget where the options are dynamically set, the same doesn't show up during Add new row flow

https://www.loom.com/share/0d9fbce9249d4445b57a231b0f0ad9e9

### Steps To Reproduce

1. Dnd Table widget
2. Change column type to select
3. Set options dynamically thru API binding
4. Click on Add new Row
5. The options doesn't show up 

### Public Sample App

_No response_

### Issue video log

_No response_

### Version

Production

Front logo [Front conversations](https://app.frontapp.com/open/top_7kj2k)

### PR DESCRIPTION
**Pull request by keyurparalkar**
## Description

This PR introduces new two new properties: `Same options in new row` and `New row options` for the select column type.
We show these two fields when the allow add new row is turned on in table 
  - SameOptionsInNewRow -- Boolean field ; When turned on the option for the first row is used in the new row. Default is true
     - Label : Use same options in new row
     - Tooltip : Toggle to display same choices for new row and editing existing row in column.
  - NewRowOptions -- Array field; when the above boolean is turned off we show this option to let people add options specifically for new row. Supports static and dynamic data, but doesn't support currentRow
    - Label: New row options
    - Tooltip : Options exclusively displayed in the column for new row addition.

Fixes #20230

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

- Cypress:
    - When allowAddNewRow is turned on, same new option should be true and new row option should be invisible.
    - When turned on same option, options of the first row should appear while editing as well as adding new row.
    - When turned off, New row option should be visible
    - Fill new row options, the new row options should appear while adding new row in select field.
    -  New row options should not have access to the currentRow
    -  Both the options should only be visible only when allowNewRow is true.
    -  Should support static and dynamic values in new row options

- Manual
    - For the new DnD Tables. Check if the above functionality is working as expected.
    - For Existing tables, Check if the above functionality is working as expected. This needed to be tested from migration standpoint

### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

### CODE REVIEW

File: app/client/src/widgets/TableWidgetV2/widget/index.tsx [hunk 1]
```diff
@@ -1441,6 +1441,21 @@ class TableWidgetV2 extends BaseWidget<TableWidgetProps, WidgetState> {
...
+      if (column.allowSameOptionsInNewRow) {
+        // Use select options from the first row
+        selectOptions = column.selectOptions[0];
+      } else {
+        selectOptions = column.newRowSelectOptions;
+      }
+    } else {
+      selectOptions = cellProperties.selectOptions;
+    }
+
```
Comment (vsvamsi1) [inline app/client/src/widgets/TableWidgetV2/widget/index.tsx:1458]: Can you extract this piece of code into a function. It seems to infer selectOptions, you could name it something like "getSelections."

File: app/client/src/utils/migrations/TableWidget.ts [hunk 2]
```diff
@@ -713,3 +714,25 @@ export const migrateColumnFreezeAttributes = (currentDSL: DSLWidget) => {
...
+
+      // Set default value for allowSameOptionsInNewRow
+      if (primaryColumns) {
+        for (const column in primaryColumns) {
+          if (
+            primaryColumns[column].hasOwnProperty("columnType") &&
+            primaryColumns[column].columnType === ColumnTypesV2.SELECT
+          ) {
+            primaryColumns[column].allowSameOptionsInNewRow = true;
+          }
```
Comment (vsvamsi1) [inline app/client/src/utils/migrations/TableWidget.ts:733]: Can you treat primaryColumns as immutable and do something like this. 
`
widget.primaryColumns = primaryColumns.map((column) => {
  if (
    column.hasOwnProperty("columnType") &&
    column.columnType === ColumnTypesV2.SELECT
  ) {
    return {
      ...column,
      allowSameOptionsInNewRow: true,
    };
  }
  return column
});
`

File: app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Widgets/TableV2/columnTypes/select_spec.js [hunk 3]
```diff
@@ -197,7 +199,99 @@ describe("Table widget - Select column type functionality", () => {
...
+    cy.get(".menu-item-text").contains("male").should("exist");
+    cy.get(".menu-item-text").contains("female").should("exist");
+    cy.get("[data-colindex=1][data-rowindex=0]").click({ force: true });
+    cy.wait(500);
+
+    // Check if the same options appears while adding a new row.
+    cy.get(".t--add-new-row").scrollIntoView().should("be.visible");
+    cy.get(".t--add-new-row").click({ force: true });
+    cy.wait(10000);
+
```
Comment (vsvamsi1) [inline app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Widgets/TableV2/columnTypes/select_spec.js:241]: Can you see if you can bring it down, 10 seconds quiet a bit. If you can't also no problem.

File: app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Widgets/TableV2/columnTypes/select_spec.js [hunk 4]
```diff
@@ -197,7 +199,99 @@ describe("Table widget - Select column type functionality", () => {
     cy.get(".menu-item-text").contains("#1").should("not.exist");
   });
 
-  it("8. should check that server side filering is working", () => {
+  it("8. should check that 'same select option in new row' property is working", () => {
```
Comment (vsvamsi1) [inline app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Widgets/TableV2/columnTypes/select_spec.js:202]: Can you add a negative test case and some comments around code blocks in the testcase explaining what that block of code is trying to do

File: app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Widgets/TableV2/columnTypes/select_spec.js (hunk 3 above)
Comment (keyurparalkar) [reply app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Widgets/TableV2/columnTypes/select_spec.js:241]: This is done. Reduced it to 500ms

File: app/client/src/widgets/TableWidgetV2/widget/utilities.ts [hunk 5]
```diff
@@ -1102,3 +1104,25 @@ export const getDragHandlers = (
...
+  },
+  cellProperties: CellLayoutProperties,
+) => {
+  if (isAddRowInProgress) {
+    if (
+      columnProperties.allowSameOptionsInNewRow &&
+      columnProperties?.selectOptions
+    ) {
+      // Use select options from the first row
+      return columnProperties.selectOptions[0] as any;
```
Comment (sbalaji1192) [inline app/client/src/widgets/TableWidgetV2/widget/utilities.ts:1121]: we need to only do this for first row not all rows.

Comment (sbalaji1192) [reply app/client/src/widgets/TableWidgetV2/widget/utilities.ts:1121]: Never use any

File: app/client/src/widgets/TableWidgetV2/widget/propertyConfig/PanelConfig/Select.ts [hunk 6]
```diff
@@ -35,6 +37,64 @@ export default {
...
+          if (
+            columnType === ColumnTypes.SELECT &&
+            props.allowAddNewRow &&
+            !allowSameOptionsInNewRow
+          ) {
+            return false;
+          }
+        }
+
+        return true;
```
Comment (sbalaji1192) [inline app/client/src/widgets/TableWidgetV2/widget/propertyConfig/PanelConfig/Select.ts:95]: do this on a else condition

File: app/client/src/widgets/TableWidgetV2/widget/index.tsx [hunk 7]
```diff
@@ -1441,6 +1442,12 @@ class TableWidgetV2 extends BaseWidget<TableWidgetProps, WidgetState> {
     const cellProperties = getCellProperties(column, originalIndex);
     let isSelected = false;
 
+    const selectOptions = getSelectOptions(
+      this.props.isAddRowInProgress,
```
Comment (sbalaji1192) [inline app/client/src/widgets/TableWidgetV2/widget/index.tsx:1446]: you need to send isNewRow

File: app/client/src/utils/migrations/TableWidget.ts [hunk 8]
```diff
@@ -713,3 +714,25 @@ export const migrateColumnFreezeAttributes = (currentDSL: DSLWidget) => {
...
+  return traverseDSLAndMigrate(currentDSL, (widget: WidgetProps) => {
+    if (widget.type === "TABLE_WIDGET_V2") {
+      const primaryColumns = widget?.primaryColumns;
+
+      // Set default value for allowSameOptionsInNewRow
+      if (primaryColumns) {
+        Object.keys(primaryColumns).forEach((column) => {
+          if (
+            primaryColumns[column].hasOwnProperty("columnType") &&
+            primaryColumns[column].columnType === ColumnTypesV2.SELECT
```
Comment (sbalaji1192) [inline app/client/src/utils/migrations/TableWidget.ts:730]: It needs to be a pure function. use constants

File: app/client/src/utils/migrations/TableWidget.ts [hunk 9]
```diff
@@ -713,3 +714,25 @@ export const migrateColumnFreezeAttributes = (currentDSL: DSLWidget) => {
...
+export const migrateSelectOptionAttributesForNewRow = (
+  currentDSL: DSLWidget,
+) => {
+  return traverseDSLAndMigrate(currentDSL, (widget: WidgetProps) => {
+    if (widget.type === "TABLE_WIDGET_V2") {
+      const primaryColumns = widget?.primaryColumns;
+
+      // Set default value for allowSameOptionsInNewRow
+      if (primaryColumns) {
+        Object.keys(primaryColumns).forEach((column) => {
```
Comment (sbalaji1192) [inline app/client/src/utils/migrations/TableWidget.ts:727]: use object.values

File: app/client/src/utils/DSLMigrations.ts [hunk 10]
```diff
@@ -1168,6 +1169,11 @@ export const transformDSL = (currentDSL: DSLWidget, newPage = false) => {
 
   if (currentDSL.version === 76) {
     currentDSL = migrateColumnFreezeAttributes(currentDSL);
+    currentDSL.version = 77;
+  }
+
+  if (currentDSL.version === 77) {
+    currentDSL = migrateSelectOptionAttributesForNewRow(currentDSL);
```
Comment (sbalaji1192) [inline app/client/src/utils/DSLMigrations.ts:1176]: table label needs to be present on the function name

**Review by sbalaji1192 — APPROVED** [review]

**Review by sbalaji1192 — APPROVED** [review]

**Comment by chandannkumar** [comment]
Tested this PR and LGTM
- Tested Add new row functionalities
- Tested Select field option visibility on adding a new row
- Tested with older apps
- Tested Migration
- Verified no breakage on migration

**Pull request merged**